{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Data cleaning"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import pandas as pd\n",
    "import os\n",
    "import bz2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "INIT_INPUT_DIR = '../metaml/data/'\n",
    "DATA_ORIGINAL_DIR = '../data-original/'\n",
    "DATA_DIR = '../data/'\n",
    "LOG_DIR = '../logs/'\n",
    "\n",
    "ABUNDANCE_FILE = 'abundance.txt'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "def exists(folder, file):\n",
    "    \"\"\"\n",
    "    Check if a file exists in a folder\n",
    "    \"\"\"\n",
    "    return os.path.exists(os.path.join(folder, file))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "def convert_bz2_txt(file_name):\n",
    "    \"\"\"\n",
    "    Opens bz2 file, reads the zipped data and writes it into the data directory\n",
    "    Returns - the path to the unzipped file\n",
    "    \"\"\"\n",
    "    input_file_path = os.path.join(INIT_INPUT_DIR, file_name)\n",
    "\n",
    "    with bz2.BZ2File(input_file_path, 'rb') as f:\n",
    "        data = f.read()\n",
    "    \n",
    "    output_file_path = os.path.join(DATA_ORIGINAL_DIR, file_name)\n",
    "\n",
    "    with open(output_file_path, 'wb') as f:\n",
    "        f.write(data)\n",
    "\n",
    "    return output_file_path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File ../data-original/abundance.txt already exists, skipping conversion...\n"
     ]
    }
   ],
   "source": [
    "abundance_file_path_og = None\n",
    "\n",
    "if(exists(DATA_ORIGINAL_DIR, ABUNDANCE_FILE)):\n",
    "    abundance_file_path_og = os.path.join(DATA_ORIGINAL_DIR, ABUNDANCE_FILE)\n",
    "    print(f\"File {abundance_file_path_og} already exists, skipping conversion...\")\n",
    "else:\n",
    "    print(\"Converting file...\")\n",
    "    abundance_file_path_og = convert_bz2_txt(ABUNDANCE_FILE + '.bz2')\n",
    "    print(\"Stored in \", abundance_file_path_og)\n",
    "\n",
    "assert abundance_file_path_og is not None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_txt(folder_name, file_name):\n",
    "    \"\"\"\n",
    "    Reads the data from the txt file\n",
    "    Returns - the data as a pandas dataframe\n",
    "    \"\"\"\n",
    "    input_file_path = os.path.join(folder_name, file_name)\n",
    "    data = pd.read_csv(input_file_path, sep='\\t', index_col=0, header=None)\n",
    "    return data\n",
    "\n",
    "def store_csv(data: pd.DataFrame, folder_name, file_name):\n",
    "    \"\"\"\n",
    "    Stores the data into a csv file in the data directory\n",
    "    Returns - the path to the csv file\n",
    "    \"\"\"\n",
    "    output_file_path = os.path.join(folder_name, file_name)\n",
    "    data.to_csv(output_file_path, index=False)\n",
    "    return output_file_path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File ../data/abundance.txt already exists, skipping conversion...\n"
     ]
    }
   ],
   "source": [
    "abundance_file_path = None\n",
    "\n",
    "if(exists(DATA_DIR, ABUNDANCE_FILE)):\n",
    "    abundance_file_path = os.path.join(DATA_DIR + ABUNDANCE_FILE)\n",
    "    print(f\"File {abundance_file_path} already exists, skipping conversion...\")\n",
    "else:\n",
    "    print(\"Converting file...\")\n",
    "    data = read_txt(DATA_ORIGINAL_DIR, ABUNDANCE_FILE)\n",
    "    data.describe()\n",
    "    data_transposed = data.transpose()\n",
    "    abundance_file_path = store_csv(data_transposed, DATA_DIR, ABUNDANCE_FILE)\n",
    "    print(\"Stored in \", abundance_file_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_657589/2935189533.py:1: DtypeWarning: Columns (2,5,9,20,26,42,47,51,52,59,61,68,76,77,78,79) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  data = pd.read_csv(abundance_file_path)\n"
     ]
    }
   ],
   "source": [
    "data = pd.read_csv(abundance_file_path)\n",
    "\n",
    "if(not exists(LOG_DIR, 'abundance_columns.txt')):\n",
    "    with open(os.path.join(LOG_DIR, 'abundance_columns.txt'), 'w') as f:\n",
    "        for col in data.columns:\n",
    "            f.write(col + '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "# First 211 columns are metadata\n",
    "metadata = data.iloc[:, :211]\n",
    "\n",
    "data_only_abundance_values = data.iloc[:, 211:]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "if(not exists(DATA_DIR, 'abundance_values.csv')):\n",
    "    print(\"Storing abundance values...\")\n",
    "    data_only_abundance_values.to_csv(os.path.join(DATA_DIR, 'abundance_values.csv'), index=False)\n",
    "\n",
    "if(not exists(DATA_DIR, 'metadata.csv')):\n",
    "    print(\"Storing metadata...\")\n",
    "    metadata.to_csv(os.path.join(DATA_DIR, 'metadata.csv'), index=False)\n",
    "\n",
    "if(not exists(LOG_DIR, \"metadata_columns.txt\")):\n",
    "    print(\"Storing metadata columns...\")\n",
    "    with open(os.path.join(LOG_DIR, \"metadata_columns.txt\"), 'w') as f:\n",
    "        for col in metadata.columns:\n",
    "            f.write(col + \", \" + metadata[col].dtype.name + '\\n')\n",
    "\n",
    "if(not exists(LOG_DIR, 'abundance_values_columns.txt')):\n",
    "    print(\"Storing abundance values columns...\")\n",
    "    with open(os.path.join(LOG_DIR, 'abundance_values_columns.txt'), 'w') as f:\n",
    "        for col in data_only_abundance_values.columns:\n",
    "            f.write(col + \", \" + data_only_abundance_values[col].dtype.name + '\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of NAN values: 0\n"
     ]
    }
   ],
   "source": [
    "# check for NAN values\n",
    "\n",
    "nan_values = data_only_abundance_values.isna().sum().sum()\n",
    "print(f\"Number of NAN values: {nan_values}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dataset_name</th>\n",
       "      <th>sampleID</th>\n",
       "      <th>subjectID</th>\n",
       "      <th>bodysite</th>\n",
       "      <th>disease</th>\n",
       "      <th>age</th>\n",
       "      <th>gender</th>\n",
       "      <th>country</th>\n",
       "      <th>sequencing_technology</th>\n",
       "      <th>pubmedid</th>\n",
       "      <th>...</th>\n",
       "      <th>statins</th>\n",
       "      <th>insulin</th>\n",
       "      <th>oral_anti-diabetic_medication</th>\n",
       "      <th>years_in_sweden</th>\n",
       "      <th>tnm_stage</th>\n",
       "      <th>ajcc_stage</th>\n",
       "      <th>localization</th>\n",
       "      <th>fobt</th>\n",
       "      <th>wif-1_gene_methylation_test</th>\n",
       "      <th>group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Candela_Africa</td>\n",
       "      <td>H10</td>\n",
       "      <td>h10</td>\n",
       "      <td>stool</td>\n",
       "      <td>n</td>\n",
       "      <td>40</td>\n",
       "      <td>female</td>\n",
       "      <td>tanzania</td>\n",
       "      <td>Illumina</td>\n",
       "      <td>25981789</td>\n",
       "      <td>...</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Candela_Africa</td>\n",
       "      <td>H11</td>\n",
       "      <td>h11</td>\n",
       "      <td>stool</td>\n",
       "      <td>n</td>\n",
       "      <td>29</td>\n",
       "      <td>female</td>\n",
       "      <td>tanzania</td>\n",
       "      <td>Illumina</td>\n",
       "      <td>25981789</td>\n",
       "      <td>...</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Candela_Africa</td>\n",
       "      <td>H12</td>\n",
       "      <td>h12</td>\n",
       "      <td>stool</td>\n",
       "      <td>n</td>\n",
       "      <td>8</td>\n",
       "      <td>female</td>\n",
       "      <td>tanzania</td>\n",
       "      <td>Illumina</td>\n",
       "      <td>25981789</td>\n",
       "      <td>...</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Candela_Africa</td>\n",
       "      <td>H13</td>\n",
       "      <td>h13</td>\n",
       "      <td>stool</td>\n",
       "      <td>n</td>\n",
       "      <td>34</td>\n",
       "      <td>male</td>\n",
       "      <td>tanzania</td>\n",
       "      <td>Illumina</td>\n",
       "      <td>25981789</td>\n",
       "      <td>...</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Candela_Africa</td>\n",
       "      <td>H14</td>\n",
       "      <td>h14</td>\n",
       "      <td>stool</td>\n",
       "      <td>n</td>\n",
       "      <td>30</td>\n",
       "      <td>male</td>\n",
       "      <td>tanzania</td>\n",
       "      <td>Illumina</td>\n",
       "      <td>25981789</td>\n",
       "      <td>...</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 211 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     dataset_name sampleID subjectID bodysite disease age  gender   country  \\\n",
       "0  Candela_Africa      H10       h10    stool       n  40  female  tanzania   \n",
       "1  Candela_Africa      H11       h11    stool       n  29  female  tanzania   \n",
       "2  Candela_Africa      H12       h12    stool       n   8  female  tanzania   \n",
       "3  Candela_Africa      H13       h13    stool       n  34    male  tanzania   \n",
       "4  Candela_Africa      H14       h14    stool       n  30    male  tanzania   \n",
       "\n",
       "  sequencing_technology  pubmedid  ... statins insulin  \\\n",
       "0              Illumina  25981789  ...      nd      nd   \n",
       "1              Illumina  25981789  ...      nd      nd   \n",
       "2              Illumina  25981789  ...      nd      nd   \n",
       "3              Illumina  25981789  ...      nd      nd   \n",
       "4              Illumina  25981789  ...      nd      nd   \n",
       "\n",
       "  oral_anti-diabetic_medication years_in_sweden tnm_stage ajcc_stage  \\\n",
       "0                            nd              nd        nd         nd   \n",
       "1                            nd              nd        nd         nd   \n",
       "2                            nd              nd        nd         nd   \n",
       "3                            nd              nd        nd         nd   \n",
       "4                            nd              nd        nd         nd   \n",
       "\n",
       "  localization fobt wif-1_gene_methylation_test group  \n",
       "0           nd   nd                          nd    nd  \n",
       "1           nd   nd                          nd    nd  \n",
       "2           nd   nd                          nd    nd  \n",
       "3           nd   nd                          nd    nd  \n",
       "4           nd   nd                          nd    nd  \n",
       "\n",
       "[5 rows x 211 columns]"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Store the describe content in a file\n",
    "\n",
    "if(not exists(LOG_DIR, 'metadata_describe.txt')):\n",
    "    with open(os.path.join(LOG_DIR, 'metadata_describe.txt'), 'w') as f:\n",
    "        # write properties of each column of metadata in a file\n",
    "        for i, col in enumerate(metadata.columns):\n",
    "            f.write(f\"Number: {i}, Column: {col}\\n\")\n",
    "            f.write(str(metadata[col].describe()) + '\\n\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unique columns: []\n"
     ]
    }
   ],
   "source": [
    "# is there any unique column in metadata?\n",
    "\n",
    "unique_columns = []\n",
    "for col in metadata.columns:\n",
    "    if len(metadata[col].unique()) == metadata.shape[0]:\n",
    "        unique_columns.append(col)\n",
    "\n",
    "print(f\"Unique columns: {unique_columns}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Candela_Africa', 'Chatelier_gut_obesity', 'doyle_bt2', 'hmp',\n",
       "       'hmpii', 'Loman2013_EcoliOutbreak_DNA_HiSeq',\n",
       "       'Loman2013_EcoliOutbreak_DNA_MiSeq', 'metahit',\n",
       "       'Neilsen_genome_assembly', 'Psoriasis_2014',\n",
       "       'Quin_gut_liver_cirrhosis', 'Segre_Human_Skin',\n",
       "       'Tito_subsistence_gut', 't2dmeta_long', 't2dmeta_short',\n",
       "       'VerticalTransmissionPilot', 'WT2D',\n",
       "       'Zeller_fecal_colorectal_cancer'], dtype=object)"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata['dataset_name'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['n', 'obesity', 'leaness', 'nd', 'stec2-positive', ' -',\n",
       "       'ibd_ulcerative_colitis', 'ibd_crohn_disease', 'n_relative', 'y',\n",
       "       'cirrhosis', 'obese', 'overweight', '-', 'underweight', 't2d',\n",
       "       'impaired_glucose_tolerance', 'cancer', 'small_adenoma',\n",
       "       'large_adenoma'], dtype=object)"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata['disease'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "# add a new column in a new dataframe called UniqueID which is a combination of dataset_name, sampleID and subjectID\n",
    "\n",
    "metadata_new = metadata.copy()\n",
    "\n",
    "# Convert all columns to string\n",
    "metadata_new['dataset_name'] = metadata_new['dataset_name'].astype(str)\n",
    "metadata_new['sampleID'] = metadata_new['sampleID'].astype(str)\n",
    "metadata_new['subjectID'] = metadata_new['subjectID'].astype(str)\n",
    "\n",
    "metadata_new['UniqueID'] = metadata_new['dataset_name'] + \"_\" +  metadata_new['sampleID'] + \"_\" + metadata_new['subjectID']\n",
    "\n",
    "# show UniqueID column in the beginning\n",
    "metadata_new = metadata_new[['UniqueID'] + [col for col in metadata_new.columns if col != 'UniqueID']]\n",
    "metadata_new.describe(include='all')\n",
    "\n",
    "# Store new metadata in a csv file\n",
    "if(not exists(DATA_DIR, 'metadata_new.csv')):\n",
    "    metadata_new.to_csv(os.path.join(DATA_DIR, 'metadata_new.csv'), index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>k__Archaea</th>\n",
       "      <th>k__Archaea|p__Euryarchaeota</th>\n",
       "      <th>k__Archaea|p__Euryarchaeota|c__Methanobacteria</th>\n",
       "      <th>k__Archaea|p__Euryarchaeota|c__Methanobacteria|o__Methanobacteriales</th>\n",
       "      <th>k__Archaea|p__Euryarchaeota|c__Methanobacteria|o__Methanobacteriales|f__Methanobacteriaceae</th>\n",
       "      <th>k__Archaea|p__Euryarchaeota|c__Methanobacteria|o__Methanobacteriales|f__Methanobacteriaceae|g__Methanobrevibacter</th>\n",
       "      <th>k__Archaea|p__Euryarchaeota|c__Methanobacteria|o__Methanobacteriales|f__Methanobacteriaceae|g__Methanobrevibacter|s__Methanobrevibacter_smithii</th>\n",
       "      <th>k__Archaea|p__Euryarchaeota|c__Methanobacteria|o__Methanobacteriales|f__Methanobacteriaceae|g__Methanobrevibacter|s__Methanobrevibacter_smithii|t__Methanobrevibacter_smithii_unclassified</th>\n",
       "      <th>k__Archaea|p__Euryarchaeota|c__Methanobacteria|o__Methanobacteriales|f__Methanobacteriaceae|g__Methanobrevibacter|s__Methanobrevibacter_unclassified</th>\n",
       "      <th>k__Archaea|p__Euryarchaeota|c__Methanobacteria|o__Methanobacteriales|f__Methanobacteriaceae|g__Methanosphaera</th>\n",
       "      <th>...</th>\n",
       "      <th>k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactobacillales|f__Enterococcaceae|g__Enterococcus|s__Enterococcus_gilvus|t__Enterococcus_gilvus_unclassified</th>\n",
       "      <th>k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactobacillales|f__Lactobacillaceae|g__Lactobacillus|s__Lactobacillus_otakiensis</th>\n",
       "      <th>k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactobacillales|f__Lactobacillaceae|g__Lactobacillus|s__Lactobacillus_otakiensis|t__GCF_000415925</th>\n",
       "      <th>k__Bacteria|p__Firmicutes|c__Clostridia|o__Clostridiales|f__Peptococcaceae</th>\n",
       "      <th>k__Bacteria|p__Firmicutes|c__Clostridia|o__Clostridiales|f__Peptococcaceae|g__Desulfotomaculum</th>\n",
       "      <th>k__Bacteria|p__Firmicutes|c__Clostridia|o__Clostridiales|f__Peptococcaceae|g__Desulfotomaculum|s__Desulfotomaculum_ruminis</th>\n",
       "      <th>k__Bacteria|p__Firmicutes|c__Clostridia|o__Clostridiales|f__Peptococcaceae|g__Desulfotomaculum|s__Desulfotomaculum_ruminis|t__GCF_000215085</th>\n",
       "      <th>k__Bacteria|p__Firmicutes|c__Clostridia|o__Clostridiales|f__Ruminococcaceae|g__Faecalibacterium|s__Faecalibacterium_prausnitzii|t__GCF_000209855</th>\n",
       "      <th>k__Bacteria|p__Firmicutes|c__Negativicutes|o__Selenomonadales|f__Veillonellaceae|g__Megasphaera|s__Megasphaera_sp_BV3C16_1</th>\n",
       "      <th>k__Bacteria|p__Firmicutes|c__Negativicutes|o__Selenomonadales|f__Veillonellaceae|g__Megasphaera|s__Megasphaera_sp_BV3C16_1|t__GCF_000478965</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.24169</td>\n",
       "      <td>0.24169</td>\n",
       "      <td>0.24169</td>\n",
       "      <td>0.24169</td>\n",
       "      <td>0.24169</td>\n",
       "      <td>0.24169</td>\n",
       "      <td>0.24169</td>\n",
       "      <td>0.24169</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.50621</td>\n",
       "      <td>0.50621</td>\n",
       "      <td>0.50621</td>\n",
       "      <td>0.50621</td>\n",
       "      <td>0.50621</td>\n",
       "      <td>0.50621</td>\n",
       "      <td>0.50621</td>\n",
       "      <td>0.50621</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.30522</td>\n",
       "      <td>0.30522</td>\n",
       "      <td>0.30522</td>\n",
       "      <td>0.30522</td>\n",
       "      <td>0.30522</td>\n",
       "      <td>0.30522</td>\n",
       "      <td>0.30522</td>\n",
       "      <td>0.30522</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.40133</td>\n",
       "      <td>0.40133</td>\n",
       "      <td>0.40133</td>\n",
       "      <td>0.40133</td>\n",
       "      <td>0.40133</td>\n",
       "      <td>0.35994</td>\n",
       "      <td>0.35994</td>\n",
       "      <td>0.35994</td>\n",
       "      <td>0.00000</td>\n",
       "      <td>0.04139</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.17479</td>\n",
       "      <td>0.17479</td>\n",
       "      <td>0.17479</td>\n",
       "      <td>0.17479</td>\n",
       "      <td>0.17479</td>\n",
       "      <td>0.16331</td>\n",
       "      <td>0.15280</td>\n",
       "      <td>0.15280</td>\n",
       "      <td>0.01051</td>\n",
       "      <td>0.01148</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 3302 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   k__Archaea  k__Archaea|p__Euryarchaeota  \\\n",
       "0     0.24169                      0.24169   \n",
       "1     0.50621                      0.50621   \n",
       "2     0.30522                      0.30522   \n",
       "3     0.40133                      0.40133   \n",
       "4     0.17479                      0.17479   \n",
       "\n",
       "   k__Archaea|p__Euryarchaeota|c__Methanobacteria  \\\n",
       "0                                         0.24169   \n",
       "1                                         0.50621   \n",
       "2                                         0.30522   \n",
       "3                                         0.40133   \n",
       "4                                         0.17479   \n",
       "\n",
       "   k__Archaea|p__Euryarchaeota|c__Methanobacteria|o__Methanobacteriales  \\\n",
       "0                                            0.24169                      \n",
       "1                                            0.50621                      \n",
       "2                                            0.30522                      \n",
       "3                                            0.40133                      \n",
       "4                                            0.17479                      \n",
       "\n",
       "   k__Archaea|p__Euryarchaeota|c__Methanobacteria|o__Methanobacteriales|f__Methanobacteriaceae  \\\n",
       "0                                            0.24169                                             \n",
       "1                                            0.50621                                             \n",
       "2                                            0.30522                                             \n",
       "3                                            0.40133                                             \n",
       "4                                            0.17479                                             \n",
       "\n",
       "   k__Archaea|p__Euryarchaeota|c__Methanobacteria|o__Methanobacteriales|f__Methanobacteriaceae|g__Methanobrevibacter  \\\n",
       "0                                            0.24169                                                                   \n",
       "1                                            0.50621                                                                   \n",
       "2                                            0.30522                                                                   \n",
       "3                                            0.35994                                                                   \n",
       "4                                            0.16331                                                                   \n",
       "\n",
       "   k__Archaea|p__Euryarchaeota|c__Methanobacteria|o__Methanobacteriales|f__Methanobacteriaceae|g__Methanobrevibacter|s__Methanobrevibacter_smithii  \\\n",
       "0                                            0.24169                                                                                                 \n",
       "1                                            0.50621                                                                                                 \n",
       "2                                            0.30522                                                                                                 \n",
       "3                                            0.35994                                                                                                 \n",
       "4                                            0.15280                                                                                                 \n",
       "\n",
       "   k__Archaea|p__Euryarchaeota|c__Methanobacteria|o__Methanobacteriales|f__Methanobacteriaceae|g__Methanobrevibacter|s__Methanobrevibacter_smithii|t__Methanobrevibacter_smithii_unclassified  \\\n",
       "0                                            0.24169                                                                                                                                            \n",
       "1                                            0.50621                                                                                                                                            \n",
       "2                                            0.30522                                                                                                                                            \n",
       "3                                            0.35994                                                                                                                                            \n",
       "4                                            0.15280                                                                                                                                            \n",
       "\n",
       "   k__Archaea|p__Euryarchaeota|c__Methanobacteria|o__Methanobacteriales|f__Methanobacteriaceae|g__Methanobrevibacter|s__Methanobrevibacter_unclassified  \\\n",
       "0                                            0.00000                                                                                                      \n",
       "1                                            0.00000                                                                                                      \n",
       "2                                            0.00000                                                                                                      \n",
       "3                                            0.00000                                                                                                      \n",
       "4                                            0.01051                                                                                                      \n",
       "\n",
       "   k__Archaea|p__Euryarchaeota|c__Methanobacteria|o__Methanobacteriales|f__Methanobacteriaceae|g__Methanosphaera  \\\n",
       "0                                            0.00000                                                               \n",
       "1                                            0.00000                                                               \n",
       "2                                            0.00000                                                               \n",
       "3                                            0.04139                                                               \n",
       "4                                            0.01148                                                               \n",
       "\n",
       "   ...  \\\n",
       "0  ...   \n",
       "1  ...   \n",
       "2  ...   \n",
       "3  ...   \n",
       "4  ...   \n",
       "\n",
       "   k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactobacillales|f__Enterococcaceae|g__Enterococcus|s__Enterococcus_gilvus|t__Enterococcus_gilvus_unclassified  \\\n",
       "0                                                0.0                                                                                                       \n",
       "1                                                0.0                                                                                                       \n",
       "2                                                0.0                                                                                                       \n",
       "3                                                0.0                                                                                                       \n",
       "4                                                0.0                                                                                                       \n",
       "\n",
       "   k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactobacillales|f__Lactobacillaceae|g__Lactobacillus|s__Lactobacillus_otakiensis  \\\n",
       "0                                                0.0                                                                          \n",
       "1                                                0.0                                                                          \n",
       "2                                                0.0                                                                          \n",
       "3                                                0.0                                                                          \n",
       "4                                                0.0                                                                          \n",
       "\n",
       "   k__Bacteria|p__Firmicutes|c__Bacilli|o__Lactobacillales|f__Lactobacillaceae|g__Lactobacillus|s__Lactobacillus_otakiensis|t__GCF_000415925  \\\n",
       "0                                                0.0                                                                                           \n",
       "1                                                0.0                                                                                           \n",
       "2                                                0.0                                                                                           \n",
       "3                                                0.0                                                                                           \n",
       "4                                                0.0                                                                                           \n",
       "\n",
       "   k__Bacteria|p__Firmicutes|c__Clostridia|o__Clostridiales|f__Peptococcaceae  \\\n",
       "0                                                0.0                            \n",
       "1                                                0.0                            \n",
       "2                                                0.0                            \n",
       "3                                                0.0                            \n",
       "4                                                0.0                            \n",
       "\n",
       "   k__Bacteria|p__Firmicutes|c__Clostridia|o__Clostridiales|f__Peptococcaceae|g__Desulfotomaculum  \\\n",
       "0                                                0.0                                                \n",
       "1                                                0.0                                                \n",
       "2                                                0.0                                                \n",
       "3                                                0.0                                                \n",
       "4                                                0.0                                                \n",
       "\n",
       "   k__Bacteria|p__Firmicutes|c__Clostridia|o__Clostridiales|f__Peptococcaceae|g__Desulfotomaculum|s__Desulfotomaculum_ruminis  \\\n",
       "0                                                0.0                                                                            \n",
       "1                                                0.0                                                                            \n",
       "2                                                0.0                                                                            \n",
       "3                                                0.0                                                                            \n",
       "4                                                0.0                                                                            \n",
       "\n",
       "   k__Bacteria|p__Firmicutes|c__Clostridia|o__Clostridiales|f__Peptococcaceae|g__Desulfotomaculum|s__Desulfotomaculum_ruminis|t__GCF_000215085  \\\n",
       "0                                                0.0                                                                                             \n",
       "1                                                0.0                                                                                             \n",
       "2                                                0.0                                                                                             \n",
       "3                                                0.0                                                                                             \n",
       "4                                                0.0                                                                                             \n",
       "\n",
       "   k__Bacteria|p__Firmicutes|c__Clostridia|o__Clostridiales|f__Ruminococcaceae|g__Faecalibacterium|s__Faecalibacterium_prausnitzii|t__GCF_000209855  \\\n",
       "0                                                0.0                                                                                                  \n",
       "1                                                0.0                                                                                                  \n",
       "2                                                0.0                                                                                                  \n",
       "3                                                0.0                                                                                                  \n",
       "4                                                0.0                                                                                                  \n",
       "\n",
       "   k__Bacteria|p__Firmicutes|c__Negativicutes|o__Selenomonadales|f__Veillonellaceae|g__Megasphaera|s__Megasphaera_sp_BV3C16_1  \\\n",
       "0                                                0.0                                                                            \n",
       "1                                                0.0                                                                            \n",
       "2                                                0.0                                                                            \n",
       "3                                                0.0                                                                            \n",
       "4                                                0.0                                                                            \n",
       "\n",
       "   k__Bacteria|p__Firmicutes|c__Negativicutes|o__Selenomonadales|f__Veillonellaceae|g__Megasphaera|s__Megasphaera_sp_BV3C16_1|t__GCF_000478965  \n",
       "0                                                0.0                                                                                            \n",
       "1                                                0.0                                                                                            \n",
       "2                                                0.0                                                                                            \n",
       "3                                                0.0                                                                                            \n",
       "4                                                0.0                                                                                            \n",
       "\n",
       "[5 rows x 3302 columns]"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_only_abundance_values.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_only_abundance_values_new = data_only_abundance_values.copy()\n",
    "data_only_abundance_values_new['UniqueID'] = metadata_new['UniqueID']\n",
    "\n",
    "# Show UniqueID column in the beginning\n",
    "data_only_abundance_values_new = data_only_abundance_values_new[['UniqueID'] + [col for col in data_only_abundance_values_new.columns if col != 'UniqueID']]\n",
    "data_only_abundance_values_new.head()\n",
    "\n",
    "# Store new data in a csv file\n",
    "if(not exists(DATA_DIR, 'abundance_values_new.csv')):\n",
    "    data_only_abundance_values_new.to_csv(os.path.join(DATA_DIR, 'abundance_values_new.csv'), index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Clean every data in memory except metadata_new and data_only_abundance_values_new\n",
    "del data\n",
    "del metadata\n",
    "del data_only_abundance_values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>UniqueID</th>\n",
       "      <th>dataset_name</th>\n",
       "      <th>sampleID</th>\n",
       "      <th>subjectID</th>\n",
       "      <th>bodysite</th>\n",
       "      <th>disease</th>\n",
       "      <th>age</th>\n",
       "      <th>gender</th>\n",
       "      <th>country</th>\n",
       "      <th>sequencing_technology</th>\n",
       "      <th>...</th>\n",
       "      <th>statins</th>\n",
       "      <th>insulin</th>\n",
       "      <th>oral_anti-diabetic_medication</th>\n",
       "      <th>years_in_sweden</th>\n",
       "      <th>tnm_stage</th>\n",
       "      <th>ajcc_stage</th>\n",
       "      <th>localization</th>\n",
       "      <th>fobt</th>\n",
       "      <th>wif-1_gene_methylation_test</th>\n",
       "      <th>group</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Candela_Africa_H10_h10</td>\n",
       "      <td>Candela_Africa</td>\n",
       "      <td>H10</td>\n",
       "      <td>h10</td>\n",
       "      <td>stool</td>\n",
       "      <td>n</td>\n",
       "      <td>40</td>\n",
       "      <td>female</td>\n",
       "      <td>tanzania</td>\n",
       "      <td>Illumina</td>\n",
       "      <td>...</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Candela_Africa_H11_h11</td>\n",
       "      <td>Candela_Africa</td>\n",
       "      <td>H11</td>\n",
       "      <td>h11</td>\n",
       "      <td>stool</td>\n",
       "      <td>n</td>\n",
       "      <td>29</td>\n",
       "      <td>female</td>\n",
       "      <td>tanzania</td>\n",
       "      <td>Illumina</td>\n",
       "      <td>...</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Candela_Africa_H12_h12</td>\n",
       "      <td>Candela_Africa</td>\n",
       "      <td>H12</td>\n",
       "      <td>h12</td>\n",
       "      <td>stool</td>\n",
       "      <td>n</td>\n",
       "      <td>8</td>\n",
       "      <td>female</td>\n",
       "      <td>tanzania</td>\n",
       "      <td>Illumina</td>\n",
       "      <td>...</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Candela_Africa_H13_h13</td>\n",
       "      <td>Candela_Africa</td>\n",
       "      <td>H13</td>\n",
       "      <td>h13</td>\n",
       "      <td>stool</td>\n",
       "      <td>n</td>\n",
       "      <td>34</td>\n",
       "      <td>male</td>\n",
       "      <td>tanzania</td>\n",
       "      <td>Illumina</td>\n",
       "      <td>...</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Candela_Africa_H14_h14</td>\n",
       "      <td>Candela_Africa</td>\n",
       "      <td>H14</td>\n",
       "      <td>h14</td>\n",
       "      <td>stool</td>\n",
       "      <td>n</td>\n",
       "      <td>30</td>\n",
       "      <td>male</td>\n",
       "      <td>tanzania</td>\n",
       "      <td>Illumina</td>\n",
       "      <td>...</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "      <td>nd</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 212 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 UniqueID    dataset_name sampleID subjectID bodysite disease  \\\n",
       "0  Candela_Africa_H10_h10  Candela_Africa      H10       h10    stool       n   \n",
       "1  Candela_Africa_H11_h11  Candela_Africa      H11       h11    stool       n   \n",
       "2  Candela_Africa_H12_h12  Candela_Africa      H12       h12    stool       n   \n",
       "3  Candela_Africa_H13_h13  Candela_Africa      H13       h13    stool       n   \n",
       "4  Candela_Africa_H14_h14  Candela_Africa      H14       h14    stool       n   \n",
       "\n",
       "  age  gender   country sequencing_technology  ... statins insulin  \\\n",
       "0  40  female  tanzania              Illumina  ...      nd      nd   \n",
       "1  29  female  tanzania              Illumina  ...      nd      nd   \n",
       "2   8  female  tanzania              Illumina  ...      nd      nd   \n",
       "3  34    male  tanzania              Illumina  ...      nd      nd   \n",
       "4  30    male  tanzania              Illumina  ...      nd      nd   \n",
       "\n",
       "  oral_anti-diabetic_medication years_in_sweden tnm_stage ajcc_stage  \\\n",
       "0                            nd              nd        nd         nd   \n",
       "1                            nd              nd        nd         nd   \n",
       "2                            nd              nd        nd         nd   \n",
       "3                            nd              nd        nd         nd   \n",
       "4                            nd              nd        nd         nd   \n",
       "\n",
       "  localization fobt wif-1_gene_methylation_test group  \n",
       "0           nd   nd                          nd    nd  \n",
       "1           nd   nd                          nd    nd  \n",
       "2           nd   nd                          nd    nd  \n",
       "3           nd   nd                          nd    nd  \n",
       "4           nd   nd                          nd    nd  \n",
       "\n",
       "[5 rows x 212 columns]"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata_new.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['stool', 'nd', 'tongue_dorsum', 'buccal_mucosa',\n",
       "       'supragingival_plaque', 'anterior_nares', 'posterior_fornix',\n",
       "       'l_retroauricular_crease', 'r_retroauricular_crease', 'saliva',\n",
       "       'keratinized_gingiva', 'palatine_tonsils', 'throat',\n",
       "       'subgingival_plaque', 'vaginal_introitus', 'mid_vagina',\n",
       "       'hard_palate', 'gomito_dx', 'gomito_sx',\n",
       "       'solco_retroauricolare_dx', 'solco_retroauricolare_sx', 'skin',\n",
       "       'mother_faeces', 'infant_faeces', 'milk'], dtype=object)"
      ]
     },
     "execution_count": 71,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata_new['bodysite'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Column: UniqueID, Type: object\n",
      "Column: dataset_name, Type: object\n",
      "Column: sampleID, Type: object\n",
      "Column: subjectID, Type: object\n",
      "Column: bodysite, Type: object\n",
      "Column: disease, Type: object\n",
      "Column: age, Type: object\n",
      "Column: gender, Type: object\n",
      "Column: country, Type: object\n",
      "Column: sequencing_technology, Type: object\n",
      "Column: pubmedid, Type: object\n",
      "Column: camp, Type: object\n",
      "Column: paired_end_insert_size, Type: object\n",
      "Column: read_length, Type: object\n",
      "Column: total_reads, Type: object\n",
      "Column: matched_reads, Type: object\n",
      "Column: uniquely_matching_reads, Type: object\n",
      "Column: uniquely_matched_reads, Type: object\n",
      "Column: gene_number, Type: object\n",
      "Column: gene_number_for_11m_uniquely_matched_reads, Type: object\n",
      "Column: hitchip_probe_number, Type: object\n",
      "Column: bmi, Type: object\n",
      "Column: gene_count_class, Type: object\n",
      "Column: hitchip_probe_class, Type: object\n",
      "Column: #SampleID, Type: object\n",
      "Column: rna_sampleid, Type: object\n",
      "Column: infant_gender, Type: object\n",
      "Column: infant_gestation_weeks, Type: object\n",
      "Column: cohort, Type: object\n",
      "Column: less_than_29weeks, Type: object\n",
      "Column: sample_collection_days, Type: object\n",
      "Column: collectionweek, Type: object\n",
      "Column: samplecollectionwindow, Type: object\n",
      "Column: gut_sample_id_ncbipublic, Type: object\n",
      "Column: gut_sample_id_corrected, Type: object\n",
      "Column: projectid, Type: object\n",
      "Column: flowcell, Type: object\n",
      "Column: comment, Type: object\n",
      "Column: mlst_project, Type: object\n",
      "Column: mlst_ec, Type: object\n",
      "Column: st_ec, Type: object\n",
      "Column: mlst_kp, Type: object\n",
      "Column: st_kp, Type: object\n",
      "Column: extractionprotocolid, Type: object\n",
      "Column: site_id_cincinnati, Type: object\n",
      "Column: delivery_mode, Type: object\n",
      "Column: infant_ethnicity, Type: object\n",
      "Column: infant_race, Type: object\n",
      "Column: birth_year, Type: object\n",
      "Column: necrotizing_enterocolitis, Type: object\n",
      "Column: dol_firstnecors, Type: object\n",
      "Column: sepsis, Type: object\n",
      "Column: infant_birthweight_kg, Type: object\n",
      "Column: infant_birth_length_cm, Type: object\n",
      "Column: maternal_abx_given, Type: object\n",
      "Column: postnatal_antimicrobial_use, Type: object\n",
      "Column: postnatal_abx_2window, Type: object\n",
      "Column: perinatal_antimicrobial_use, Type: object\n",
      "Column: death, Type: object\n",
      "Column: nursing_status, Type: object\n",
      "Column: maternal_age_at_delivery_years, Type: object\n",
      "Column: mat_age_bin, Type: object\n",
      "Column: parity, Type: object\n",
      "Column: infant_fut2, Type: object\n",
      "Column: mat_fut2, Type: object\n",
      "Column: lowh, Type: object\n",
      "Column: lowlea, Type: object\n",
      "Column: highslea, Type: object\n",
      "Column: married, Type: object\n",
      "Column: gravida, Type: object\n",
      "Column: preeclampsia, Type: object\n",
      "Column: mult_birth, Type: object\n",
      "Column: hypertension, Type: object\n",
      "Column: hypertension_prepreg, Type: object\n",
      "Column: chorioamnionitis, Type: object\n",
      "Column: primipar, Type: object\n",
      "Column: daysonform3daysprior, Type: object\n",
      "Column: days_on_abx_14, Type: object\n",
      "Column: visit_number, Type: object\n",
      "Column: snprnt, Type: object\n",
      "Column: wmsphase, Type: object\n",
      "Column: first, Type: object\n",
      "Column: repeat, Type: object\n",
      "Column: stooltexture, Type: object\n",
      "Column: daysafteronset, Type: object\n",
      "Column: hus, Type: object\n",
      "Column: stec_count, Type: object\n",
      "Column: shigatoxin2elisa, Type: object\n",
      "Column: readsmillions, Type: object\n",
      "Column: nonhuman, Type: object\n",
      "Column: stec_coverage, Type: object\n",
      "Column: stxab_detected, Type: object\n",
      "Column: stx_ratio, Type: object\n",
      "Column: typingdata, Type: object\n",
      "Column: c_difficile_frequency, Type: object\n",
      "Column: ibd, Type: object\n",
      "Column: sampling_day, Type: object\n",
      "Column: known_consumers_of_a_defined_fermented_milk_product_(dfmp), Type: object\n",
      "Column: mgs_richness, Type: object\n",
      "Column: mgs_profile_matched_sample_pairs, Type: object\n",
      "Column: ethnicity, Type: object\n",
      "Column: sample_name, Type: object\n",
      "Column: index_2_primer, Type: object\n",
      "Column: index_1_primer, Type: object\n",
      "Column: sample_name_repeat1, Type: object\n",
      "Column: extraction_conc, Type: object\n",
      "Column: sampleid_with_barcode, Type: object\n",
      "Column: initial_nreads, Type: object\n",
      "Column: quality_control, Type: object\n",
      "Column: screensensloc_hg19, Type: object\n",
      "Column: screensensloc_phix174, Type: object\n",
      "Column: screenhiv, Type: object\n",
      "Column: percent_initial_reads, Type: object\n",
      "Column: percent_qc_reads, Type: object\n",
      "Column: tipo_prelievo, Type: object\n",
      "Column: prima_dopo_cura, Type: object\n",
      "Column: codice_provetta, Type: object\n",
      "Column: data, Type: object\n",
      "Column: prelievo_efemalefemaleettuato_presso, Type: object\n",
      "Column: patient_number, Type: object\n",
      "Column: pasi, Type: object\n",
      "Column: bsa, Type: object\n",
      "Column: age_ofemale_onset, Type: object\n",
      "Column: arthritis, Type: object\n",
      "Column: type, Type: object\n",
      "Column: antibiotic_usage, Type: object\n",
      "Column: stage, Type: object\n",
      "Column: cirrhotic, Type: object\n",
      "Column: hbv_related, Type: object\n",
      "Column: alcohol_related, Type: object\n",
      "Column: other_causes_related, Type: object\n",
      "Column: inr, Type: object\n",
      "Column: crea, Type: object\n",
      "Column: alb, Type: object\n",
      "Column: tb, Type: object\n",
      "Column: pt, Type: object\n",
      "Column: ascites, Type: object\n",
      "Column: he, Type: object\n",
      "Column: ctp, Type: object\n",
      "Column: meld, Type: object\n",
      "Column: antivirus, Type: object\n",
      "Column: ?-blocker, Type: object\n",
      "Column: designation, Type: object\n",
      "Column: age_range, Type: object\n",
      "Column: site_symmetry, Type: object\n",
      "Column: affected/unaffected, Type: object\n",
      "Column: method, Type: object\n",
      "Column: site_characteristic, Type: object\n",
      "Column: total_intital_reads, Type: object\n",
      "Column: estimated_median_insert_size, Type: object\n",
      "Column: reported_as_failed_qc, Type: object\n",
      "Column: uniquely_align_to_human, Type: object\n",
      "Column: non_uniquely_align_to_human_with_0_2_mismatches, Type: object\n",
      "Column: reads_removed_because_of_read_pair_trimming_discrepancy, Type: object\n",
      "Column: too_short_after_quality_trimming(<50bp), Type: object\n",
      "Column: final_#_paired_reads, Type: object\n",
      "Column: final_#_singletons, Type: object\n",
      "Column: total_final_reads, Type: object\n",
      "Column: %_reads_kept_by_mean_slope_of_last_5_points_(coverage), Type: object\n",
      "Column: population, Type: object\n",
      "Column: bmi_class, Type: object\n",
      "Column: 16s_rrna, Type: object\n",
      "Column: shotgun_metagenome, Type: object\n",
      "Column: read_depth_16s, Type: object\n",
      "Column: paired, Type: object\n",
      "Column: single, Type: object\n",
      "Column: height, Type: object\n",
      "Column: weight, Type: object\n",
      "Column: diabetic, Type: object\n",
      "Column: fbg, Type: object\n",
      "Column: sbp, Type: object\n",
      "Column: dbp, Type: object\n",
      "Column: fins, Type: object\n",
      "Column: fcp, Type: object\n",
      "Column: hbalc, Type: object\n",
      "Column: tg, Type: object\n",
      "Column: tcho, Type: object\n",
      "Column: hdl, Type: object\n",
      "Column: ldl, Type: object\n",
      "Column: kit, Type: object\n",
      "Column: sample_group, Type: object\n",
      "Column: illumina_run, Type: object\n",
      "Column: classification, Type: object\n",
      "Column: gad-antibodies, Type: object\n",
      "Column: whr, Type: object\n",
      "Column: wc, Type: object\n",
      "Column: cholesterol, Type: object\n",
      "Column: triglycerides, Type: object\n",
      "Column: creatinine, Type: object\n",
      "Column: y-gt, Type: object\n",
      "Column: fasting_glucose, Type: object\n",
      "Column: fasting_insulin, Type: object\n",
      "Column: hba1c, Type: object\n",
      "Column: adiponectin, Type: object\n",
      "Column: leptin, Type: object\n",
      "Column: glp-1, Type: object\n",
      "Column: fgf-19, Type: object\n",
      "Column: hscrp, Type: object\n",
      "Column: c-peptide, Type: object\n",
      "Column: tnfa, Type: object\n",
      "Column: il-1, Type: object\n",
      "Column: cd163, Type: object\n",
      "Column: statins, Type: object\n",
      "Column: insulin, Type: object\n",
      "Column: oral_anti-diabetic_medication, Type: object\n",
      "Column: years_in_sweden, Type: object\n",
      "Column: tnm_stage, Type: object\n",
      "Column: ajcc_stage, Type: object\n",
      "Column: localization, Type: object\n",
      "Column: fobt, Type: object\n",
      "Column: wif-1_gene_methylation_test, Type: object\n",
      "Column: group, Type: object\n"
     ]
    }
   ],
   "source": [
    "for col in metadata_new.columns:\n",
    "    print(f\"Column: {col}, Type: {metadata_new[col].dtype.name}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
